compiler/src/main/java/com/squareup/anvil/compiler/codegen/reference/AnnotationReferenceIr.kt

```diff
@@ -0,0 +1,31 @@
+package com.squareup.anvil.compiler.codegen.reference
+
+import com.squareup.anvil.compiler.api.AnvilCompilationException
+import com.squareup.anvil.compiler.fqName
+import com.squareup.anvil.compiler.scope
+import org.jetbrains.kotlin.backend.common.extensions.IrPluginContext
+import org.jetbrains.kotlin.ir.expressions.IrConstructorCall
+import org.jetbrains.kotlin.ir.util.parentAsClass
+import org.jetbrains.kotlin.name.FqName
+import kotlin.LazyThreadSafetyMode.NONE
+
+internal class AnnotationReferenceIr(
+  val annotation: IrConstructorCall,
+  val context: IrPluginContext
+) {
+  val fqName: FqName = annotation.symbol.owner.parentAsClass.fqName
+
+  val scope: ClassReferenceIr by lazy(NONE) {
+    scopeOrNull ?: throw AnvilCompilationException(
+      element = annotation,
+      message = "Couldn't find scope for $fqName."
+    )
+  }
+
+  val scopeOrNull: ClassReferenceIr? by lazy(NONE) {
+    context.referenceClass(annotation.scope())?.toClassReference(context)
+  }
```

Comment:
Same here for `toString()`, `hashCode()` and `equals()` (hashCode and equals might not work yet without parsing annotation arguments).